doc/proposal/monitoring.md

```diff
@@ -0,0 +1,144 @@
+# Design Proposal: Monitoring the Istio Service Mesh
+
+Status: **DRAFT**
+
+- [Introduction](#Introduction)
+- [Proposed Design](#Proposed_Design)
+  - [Design Summary](#Design_Summary)
+  - [Telemetry Generation And Reporting](#Telemetry_Generation_And_Reporting)
+    - [Schema Publishing](#Schema_Publishing)
+    - [Example Library Pseudocode](#Example_Library_Pseudocode)
+  - [Mixer-Specific Telemetry](#Mixer-Specific_Telemetry)
+
+## Introduction
+
+The [Istio Service Mesh](http://github.com/istio/istio) provides automated 
+telemetry data collection for integrated services. As requests flow through the
+service mesh, request metadata is generated by the [Proxy](http://github.com/istio/proxy) 
+and sent to the [Mixer](http://github.com/istio/mixer). The Mixer repackages the
+received metadata, generating access logs and request metrics, and transmits the
+telemetry data to a configurable set of backend systems for monitoring and 
+alerting.
+
+Istio development has, so far, focused on providing the subsystems for the 
+reporting of telemetry data and for tailoring the collection to match the intent
+of the service operators. Little attention has been devoted to monitoring the 
+behavior of the individual Istio components (Proxy, [Manager](http://github.com/istio/manager),
+Mixer).
+
+This document proposes a high-level design for the monitoring of the Istio 
+Service Mesh itself.
+
+## Proposed Design
+
+The proposed design adopts two high-level goals:
+1. **Limit special handling for Istio components**
+   - Telemetry data for the Istio Proxy and Istio Manager components _should_ be
+     reported via the mixer through the same mechanisms as telemetry data for 
+     the non-Istio services within the service mesh.
+   - Telemetry data for the Istio Mixer itself _should_ be sent through the same 
```

Comment:
Why is this a requirement? In particular I'm worried about the case where the mixer is misconfigured and cannot report metrics/logs - how do we detect and report that case? As you point out below, it's not necessarily the common case that metrics are reported through Chemist internally. Certainly none of the services I've worked on internally reported metrics via Chemist. Why impose this as a requirement for Istio components, and the mixer especially (which is even more a special case than the other Istio components - quis custodiet ipsos custodes)?

Consider the following straw man instead:
As part of installing Istio in your cluster a Istio-private statsd<sup>[1](#footnote1)</sup> + datastore instances are stood up. Critical metrics (received requests, response status by code, latency per method, number of outstanding requests, amount of outstanding metrics/logs to report, etc.) are pushed to this statsd instance. These metrics are sent to statsd via a normal statsd client, not touching any reporting code paths in the mixer. We use those metrics to ensure the deployment is healthy.

The obvious downside is that under the scheme you propose, metrics for Istio components show up next to metrics for my own services, while under this straw man I have to go somewhere else to look at the health of my Istio deployment. This could be remedied by double-reporting mixer metrics: metrics are reported to statsd out of band of any mixer report-specific codepaths, and _also_ reported via a self `Report` so they're exported to whatever metrics backends I have configured.

<hr>
<a name="footnote1">1.</a> It could be Prometheus or whatever other stats backend we want. I propose statsd in this example because Envoy already self reports via statsd, which means we can easily extend this straw man to the proxy reporting critical metrics out of band of the mixer too. This is especially important if the mixer is running behind a proxy itself.